### PR TITLE
Moved onLivingDamagePre to be before absorption logic

### LINKED ISSUE #1267: Move `LivingDamageEvent.Pre` to be calculated before absorption
As discussed in discord: https://discord.com/channels/313125603924639766/1198800008225501285/1260054539084497027

It makes more sense to consider absorption as part of health, rather than a kind of damage reduction. Hence, it's better to move `LivingDamageEvent.Pre` calculation to be before absorption calculation.

Example:
I have a defensive gear that reduce damage by 50% and I want it to be implemented before magic and armor, because it could have some cost related to it. I takes 10 damage:
- If I have 10 absorption, current system will consume all 10 absorption, but it should consume 5 only.
- If I have 4 absorption, current system will deal me 3 damage, but it should deal me 1 damage only.

### PR DESCRIPTION
**Pull request by TelepathicGrunt**
The idea is that all damage manipulations should be done before absorption. Ideally absorption only absorbs the final damage amount and any remaining damage is applied to player's normal health.

Discord discussion:  https://discord.com/channels/313125603924639766/1198800008225501285/1260054539084497027 

Closes https://github.com/neoforged/NeoForge/issues/1267

### CODE REVIEW

File: patches/net/minecraft/world/entity/LivingEntity.java.patch [hunk 1]
```diff
@@ -414,11 +414,11 @@
...
 +            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ARMOR, this.damageContainers.peek().getNewDamage() - this.getDamageAfterArmorAbsorb(p_21240_, this.damageContainers.peek().getNewDamage()));
 +            this.getDamageAfterMagicAbsorb(p_21240_, this.damageContainers.peek().getNewDamage());
-+            float damage = this.damageContainers.peek().getNewDamage();
++            float damage = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
 +            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION, Math.min(this.getAbsorptionAmount(), damage));
 +            float absorbed = Math.min(damage, this.damageContainers.peek().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION));
 +            this.setAbsorptionAmount(Math.max(0, this.getAbsorptionAmount() - absorbed));
-+            float f1 = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
++            float f1 = damage - absorbed;
```
Comment (Caltinor) [inline patches/net/minecraft/world/entity/LivingEntity.java.patch]: you don't need to calculate this here.  just pass in `this.damageContainers.peek().getNewDamage()`  the `setReduction` call 3 lines up already subtracts the containers newDamage value by the reduction

File: patches/net/minecraft/world/entity/player/Player.java.patch [hunk 2]
```diff
@@ -162,11 +162,11 @@
...
 +            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ARMOR, this.damageContainers.peek().getNewDamage() - this.getDamageAfterArmorAbsorb(p_36312_, this.damageContainers.peek().getNewDamage()));
 +            this.getDamageAfterMagicAbsorb(p_36312_, this.damageContainers.peek().getNewDamage());
-+            float damage = this.damageContainers.peek().getNewDamage();
++            float damage = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
 +            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION, Math.min(this.getAbsorptionAmount(), damage));
 +            float absorbed = Math.min(damage, this.damageContainers.peek().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION));
 +            this.setAbsorptionAmount(Math.max(0, this.getAbsorptionAmount() - absorbed));
-+            float f1 = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
++            float f1 = damage - absorbed;
```
Comment (Caltinor) [inline patches/net/minecraft/world/entity/player/Player.java.patch]: same as with LivingEntity.

File: tests/src/main/java/net/neoforged/neoforge/debug/entity/living/LivingEntityEventTests.java [hunk 3]
```diff
@@ -337,6 +339,13 @@ static void livingDamageSequenceEvents(final DynamicTest test, final Registratio
         /* This event captures the change in new damage from the previous event for use in checks.*/
         test.eventListeners().forge().addListener((final LivingDamageEvent.Post event) -> {
             if (event.getEntity() instanceof GameTestPlayer player && Objects.equals(player.getCustomName(), NAME)) {
+                try {
+                    MethodHandle methodHandle = ReflectionUtils.fieldHandle(LivingEntity.class.getDeclaredField("damageContainers"));
```
Comment (Caltinor) [inline tests/src/main/java/net/neoforged/neoforge/debug/entity/living/LivingEntityEventTests.java]: Why reflection and not just `event.getReduction(Reduction.ABSORPTION)`?  The event is passed in the damage container's values during construction only to ensure they remain immutable.

File: patches/net/minecraft/world/entity/LivingEntity.java.patch (hunk 1 above)
Comment (TelepathicGrunt) [reply patches/net/minecraft/world/entity/LivingEntity.java.patch:421]: This post event doesn’t get the damage container. I checked

File: tests/src/main/java/net/neoforged/neoforge/debug/entity/living/LivingEntityEventTests.java (hunk 3 above)
Comment (TelepathicGrunt) [reply tests/src/main/java/net/neoforged/neoforge/debug/entity/living/LivingEntityEventTests.java]: This post event doesn’t have a getter for the damage container. I checked. Unless I missed it when checking. I’ll look again

Comment (Caltinor) [reply tests/src/main/java/net/neoforged/neoforge/debug/entity/living/LivingEntityEventTests.java]: you don't need the container.  this method is all you need  https://github.com/neoforged/NeoForge/blob/d71d4e59e9e1ef410f7d5284a5b3126c358bcdf7/src/main/java/net/neoforged/neoforge/event/entity/living/LivingDamageEvent.java#L153-L155

**Comment by KnightMiner** [comment]
I have not been keeping up with damage PRs lately, with this change would there still be a way to detect that the player's health actually got reduced (not absorption)? Don't care if its before or after taking damage for my usecase. I'm guessing by the name "pre" there is an event that fires post damage.

I don't have much opinion either way about whether absorption loss should be considered part of health or not, its probably cleaner if it is and certainty would have made my life easier in the past, though the new damage container API should hopefully solve my issues with a simpler implementation.

**Comment by TelepathicGrunt** [comment]
There is a LivingDamageEvent.Post event that fires after the entity is damaged. As for determining how much damage the player took, @Caltinor would have better insight than me on how that would look for your use case

**Comment by KnightMiner** [comment]
Damage container should have that info in it. I've just not looked at the damage container rework outside of the proposal yet.

**Comment by Caltinor** [comment]
> Damage container should have that info in it. I've just not looked at the damage container rework outside of the proposal yet.

`DamageContainer` will have that information up through `LivingDamageEvent.Pre`.  Once we get to `LivingDamageEvent.Post`, though, we are in an immutable state, so the event fields represent an immutable clone of what is/was in the damage container.  

For your original question, Tele's response is spot on.  You would want to use `Post#getNewDamage` to get the amount that the health was reduced by.  The change this PR is resolving is the fact that health and absorption are resources that can be depleted, unlike effects, armor, and enchantments which do not have an upper limit to their ability to reduce damage.  As such, we wanted a way to waterfall down the damage value across these depletable resources in a way that wouldn't allow for further modification that would create inconsistent behavior

**Review by Caltinor — APPROVED** [review]

**Review by Matyrobbrt — APPROVED** [review]

**Pull request merged**